[FSDP] Start to generalize modules to ignore for mixed precision

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #102231
* #102043
* #102040
* __->__ #102010

The main use case here is that folks would like to ignore layer norm for mixed precision. This can now be enabled with:

```
mp_config = MixedPrecision(
            param_dtype=torch.float16,
            reduce_dtype=torch.float16,
            buffer_dtype=torch.float16,
            _mixed_precision_module_classes_to_ignore=[_BatchNorm, nn.LayerNorm],
        )
```

This is done by classes of types in `_mixed_precision_module_classes_to_ignore` being wrapped in their own FSDP unit with mixed preicsion disabled. This is only enabled for auto wrapping.

We also add module pre and post hooks to cast / downcast inputs to the appropriate full precision.

Differential Revision: [D46079957](https://our.internmc.facebook.com/intern/diff/D46079957/)

rohan-varma committed:
[FSDP] Start to generalize modules to ignore for mixed precision

The main use case here is that folks would like to ignore layer norm for mixed precision. This can now be enabled with:

```
mp_config = MixedPrecision(
            param_dtype=torch.float16,
            reduce_dtype=torch.float16,
            buffer_dtype=torch.float16,
            _mixed_precision_module_classes_to_ignore=[_BatchNorm, nn.LayerNorm],
        )
```

for LayerNorm, we also add module pre and post hooks to cast inputs to the appropriate full precision. This is not super generalizable at the moment, for example, if users want to skip module X, we won't install the same hooks for X, but we perhaps should. This can be done as follow up work for now as the main priority is to unblock layer norm skipping.

Differential Revision: [D46079957](https://our.internmc.facebook.com/intern/diff/D46079957/)

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/_utils.py:
```diff
@@ -2,13 +2,47 @@
+def _override_module_mixed_precision(root, module_to_override, wrap_override_dict={"mixed_precision": None}):
+    for mod in root.modules():
+        if isinstance(mod, module_to_override):
+            mod._wrap_overrides = wrap_override_dict  # type: ignore[assignment]
```
This looks unused?

rohan-varma replied:
wrap_overrides is populated here and then used in auto wrapping: https://fburl.com/code/7ypnkd83

awgu commented on torch/distributed/fsdp/_utils.py:
```diff
@@ -2,13 +2,47 @@
+        if isinstance(mod, module_to_override):
+            mod._wrap_overrides = wrap_override_dict  # type: ignore[assignment]
+
+        if isinstance(mod, LayerNorm): # TODO: generalize this logic if additional types need to be supported
```
I think this should be replaced with `if isinstance(mod, module_to_override):`?

rohan-varma replied:
I was hardcoding this to LN for now. LN would crash if it got fp16 inputs and was running in fp32, but for some reason batchnorm doesn't need this. Maybe we should just always cast up to fp32?

awgu replied:
Oh I see. If batch norm kernels have the casting logic, then maybe layer norm should as well? It might be good to check on why there is that difference before adding logic at the FSDP level.

awgu commented on torch/distributed/fsdp/api.py:
```diff
@@ -133,6 +134,11 @@ class MixedPrecision:
             It takes precedence over ``cast_forward_inputs`` for the root FSDP instance.
             (Default: ``True``)
+        _mixed_precision_module_classes_to_ignore: (Sequence[type]): Module classes to ignore
+            for mixed precision. This will make the specified nn.Module types ignore mixed precision,
```
I think this comment should mention that this setting only affects auto wrapping.

awgu commented on torch/distributed/fsdp/_wrap_utils.py:
```diff
@@ -57,10 +62,21 @@ def _auto_wrap(
+        _contains_module(root_module, module_to_override)
+        for module_to_override in mixed_precision._mixed_precision_module_classes_to_ignore
+    ):
+        for mp_module_to_override in mixed_precision._mixed_precision_module_classes_to_ignore:
```
Is one of these for loops redundant?

rohan-varma replied:
yeah

awgu commented on torch/distributed/fsdp/_utils.py:
```diff
@@ -2,13 +2,47 @@
+                    return _apply_to_tensors(partial(cast_fn, old_dtype), output)
+
+            mod.register_forward_pre_hook(ln_forward_pre_hook)
+            mod.register_forward_hook(ln_forward_post_hook)
```
Do we strictly need these hooks for correctness? If so, could we make sure we have unit tests that exercise that?

rohan-varma replied:
Yeah, with the LN added to the test, it would crash without the pre hook. The necesity of the post hook depends on the `old_dtype` discussion above

awgu reviewed (commented):
I think the general idea makes sense to me, but there is some work to clean up the PR before landing. I am mostly interested in what is the minimal set of changes as there seems to be some redundancy.

It seems like there are two things going on: the added hooks for non-FSDP submodules within an FSDP's tree and also auto-wrapping specified module types in their own FSDP instances. The PR summary focuses on the former -- should we also mention the latter if we include it in this PR?

rohan-varma commented on torch/distributed/fsdp/_utils.py:
```diff
@@ -2,13 +2,47 @@
+
+        if isinstance(mod, LayerNorm): # TODO: generalize this logic if additional types need to be supported
+
+            old_dtype = None
```
Wonder what we can do here. The input -> output mapping for a module is not necessarily 1:1. We're basically saving an `old_dtype` so that we cast the output back to whatever the input precision was before forcing to fp32. But agreed there can be multiple input precisions.

Maybe we don't cast back at all (so we don't need a post hook or old dtype), and rely on users doing `cast_forward_inputs` configuring correctly to ensure dtypes into subsequent mixed precision modules are correct?

rohan-varma commented on torch/distributed/fsdp/_utils.py:
```diff
@@ -2,13 +2,47 @@
+                if old_dtype is not None:
+                    return _apply_to_tensors(partial(cast_fn, old_dtype), output)
+
+            mod.register_forward_pre_hook(ln_forward_pre_hook)
```
done

rohan-varma committed:
Update on "[FSDP] Start to generalize modules to ignore for mixed precision"


The main use case here is that folks would like to ignore layer norm for mixed precision. This can now be enabled with:

```
mp_config = MixedPrecision(
            param_dtype=torch.float16,
            reduce_dtype=torch.float16,
            buffer_dtype=torch.float16,
            _mixed_precision_module_classes_to_ignore=[_BatchNorm, nn.LayerNorm],
        )
```

for LayerNorm, we also add module pre and post hooks to cast inputs to the appropriate full precision. This is not super generalizable at the moment, for example, if users want to skip module X, we won't install the same hooks for X, but we perhaps should. This can be done as follow up work for now as the main priority is to unblock layer norm skipping.

Differential Revision: [D46079957](https://our.internmc.facebook.com/intern/diff/D46079957/)

[ghstack-poisoned]

awgu reviewed (commented):
I sent a message on chat to discuss more about the hooks casting.

rohan-varma committed:
Update on "[FSDP] Start to generalize modules to ignore for mixed precision"


The main use case here is that folks would like to ignore layer norm for mixed precision. This can now be enabled with:

```
mp_config = MixedPrecision(
            param_dtype=torch.float16,
            reduce_dtype=torch.float16,
            buffer_dtype=torch.float16,
            _mixed_precision_module_classes_to_ignore=[_BatchNorm, nn.LayerNorm],
        )
```

This is done by classes of types in `_mixed_precision_module_classes_to_ignore` being wrapped in their own FSDP unit with mixed preicsion disabled. This is only enabled for auto wrapping.

We also add module pre and post hooks to cast / downcast inputs to the appropriate full precision.

Differential Revision: [D46079957](https://our.internmc.facebook.com/intern/diff/D46079957/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] Start to generalize modules to ignore for mixed precision"


The main use case here is that folks would like to ignore layer norm for mixed precision. This can now be enabled with:

```
mp_config = MixedPrecision(
            param_dtype=torch.float16,
            reduce_dtype=torch.float16,
            buffer_dtype=torch.float16,
            _mixed_precision_module_classes_to_ignore=[_BatchNorm, nn.LayerNorm],
        )
```

This is done by classes of types in `_mixed_precision_module_classes_to_ignore` being wrapped in their own FSDP unit with mixed preicsion disabled. This is only enabled for auto wrapping.

We also add module pre and post hooks to cast / downcast inputs to the appropriate full precision.

Differential Revision: [D46079957](https://our.internmc.facebook.com/intern/diff/D46079957/)

[ghstack-poisoned]

awgu reviewed (approved):
This makes sense to me! It will be good to think more about the hooks design as we try this out in more use cases.

awgu commented:
Lint error looks real :/

rohan-varma committed:
Update on "[FSDP] Start to generalize modules to ignore for mixed precision"


The main use case here is that folks would like to ignore layer norm for mixed precision. This can now be enabled with:

```
mp_config = MixedPrecision(
            param_dtype=torch.float16,
            reduce_dtype=torch.float16,
            buffer_dtype=torch.float16,
            _mixed_precision_module_classes_to_ignore=[_BatchNorm, nn.LayerNorm],
        )
```

This is done by classes of types in `_mixed_precision_module_classes_to_ignore` being wrapped in their own FSDP unit with mixed preicsion disabled. This is only enabled for auto wrapping.

We also add module pre and post hooks to cast / downcast inputs to the appropriate full precision.

Differential Revision: [D46079957](https://our.internmc.facebook.com/intern/diff/D46079957/)

[ghstack-poisoned]

rohan-varma commented:
@pytorchbot merge -f "CI done"
